CMakeLists.txt

```diff
@@ -6,6 +6,7 @@ cmake_minimum_required(VERSION 3.20.0)
 project(onnx-mlir)
 
 option(ONNX_MLIR_BUILD_TESTS "Build ONNX-MLIR test executables. If OFF, just generate build targets." ON)
+option(ONNX_MLIR_BUILD_SHARED "Enable ONNX-MLIR shared library object." OFF)
```

Comment:
> Building static or shared onnx-mlir is always in sync with whether llvm-project was built static or shared. What's the purpose of this option to force onnx-mlir to be built differently from llvm-project?

* YES: Regardless that llvm-project (or the optional ```BUILD_SHARED_LIBS``` override) is saying  the ```libOM*.{a,so}``` objects will be generated, just as-like before.
* NO: This ```ONNX_MLIR_BUILD_SHARED``` **additionaly**  will generate a SHARED unified object.

This option here generates a single shared object **regardless** of decupled of what llvm-project is saying (or what  ```BUILD_SHARED_LIBS``` override is saying)

---

To sum up, in the end you can have these:

  * All the many previous ```libOM*.a``` static objects (plus/minus a grand ```libONNXMLIR.so```)
  * All the many previous ```libOM*.so``` shared objects (plus/minus a grand ```libONNXMLIR.so```)
 

---

Let's take LLVM and MLIR, they also ship (additional !) unified objects like ```libMLIR.so``` / ```libLLVM.so```
```
$ rpm -ql llvm18-devel | grep '\.so'
/usr/lib64/llvm18/lib/libLLVM.so
/usr/lib64/llvm18/lib/libLTO.so
[cbalint@yoda SANDBOX]$ rpm -ql llvm18-static | grep '\.a'
/usr/lib64/llvm18/lib/libLLVMAArch64AsmParser.a
/usr/lib64/llvm18/lib/libLLVMAArch64CodeGen.a
/usr/lib64/llvm18/lib/libLLVMAArch64Desc.a
/usr/lib64/llvm18/lib/libLLVMAArch64Disassembler.a
{...}

$ rpm -ql llvm18-static | grep '\.a'
/usr/lib64/llvm18/lib/libLLVMAArch64AsmParser.a
/usr/lib64/llvm18/lib/libLLVMAArch64CodeGen.a
/usr/lib64/llvm18/lib/libLLVMAArch64Desc.a
/usr/lib64/llvm18/lib/libLLVMAArch64Disassembler.a
/usr/lib64/llvm18/lib/libLLVMAArch64Info.a
/usr/lib64/llvm18/lib/libLLVMAArch64Utils.a
/usr/lib64/llvm18/lib/libLLVMAMDGPUAsmParser.a
/usr/lib64/llvm18/lib/libLLVMAMDGPUCodeGen.a
/usr/lib64/llvm18/lib/libLLVMAMDGPUDesc.a
```

```
$ rpm -ql mlir18-devel | grep '\.so'
/usr/lib64/llvm18/lib/libMLIR-C.so
/usr/lib64/llvm18/lib/libMLIR.so
/usr/lib64/llvm18/lib/libmlir_arm_sme_abi_stubs.so
/usr/lib64/llvm18/lib/libmlir_async_runtime.so
/usr/lib64/llvm18/lib/libmlir_c_runner_utils.so
/usr/lib64/llvm18/lib/libmlir_float16_utils.so
/usr/lib64/llvm18/lib/libmlir_runner_utils.so
{...}

$ rpm -ql mlir18-static | grep '\.a'
/usr/lib64/llvm18/lib/libMLIRAMDGPUDialect.a
/usr/lib64/llvm18/lib/libMLIRAMDGPUToROCDL.a
/usr/lib64/llvm18/lib/libMLIRAMDGPUTransforms.a
/usr/lib64/llvm18/lib/libMLIRAMDGPUUtils.a
/usr/lib64/llvm18/lib/libMLIRAMXDialect.a
/usr/lib64/llvm18/lib/libMLIRAMXToLLVMIRTranslation.a
/usr/lib64/llvm18/lib/libMLIRAMXTransforms.a

```

Reply:
For that purpose, the name `ONNX_MLIR_BUILD_SHARED` is a bit confusing because typically YES means a shared build while NO means a static build. `ONNX_MLIR_BUILD_SINGLE` or `ONNX_MLIR_BUILD_UNIFIED` would be a better name.

Reply:
> For that purpose, the name `ONNX_MLIR_BUILD_SHARED` is a bit confusing because typically YES means a shared build while NO means a static build. `ONNX_MLIR_BUILD_SINGLE` or `ONNX_MLIR_BUILD_UNIFIED` would be a better name.

Yes I agree, it may be confusing, any suggestion for this ?
  * My one would be: s/ONNX_MLIR_BUILD_SHARED/ONNX_MLIR_BUILD_SINGLE_SHARED/
  * But again, one who use these "extra" switches really knows what is to be expected.

Reply:
> > For that purpose, the name `ONNX_MLIR_BUILD_SHARED` is a bit confusing because typically YES means a shared build while NO means a static build. `ONNX_MLIR_BUILD_SINGLE` or `ONNX_MLIR_BUILD_UNIFIED` would be a better name.
> 
> Yes I agree, it may be confusing, any suggestion for this ?
> 
> * My one would be: s/ONNX_MLIR_BUILD_SHARED/ONNX_MLIR_BUILD_SINGLE_SHARED/
> * But again, one who use these "extra" switches really knows what is to be expected.

In an attempt to better mark this, I added (pushed now) the ["additional" word](https://github.com/onnx/onnx-mlir/pull/2700/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR9) to the very cmake option list.

```
$ cat CMakeLists.txt | grep additional
option(ONNX_MLIR_BUILD_SHARED "Enable ONNX-MLIR additional shared library object." OFF)
```

Reply:
>I think adding the word "additional" actually makes it more confusing because for most people the meaning of "additional shared library object" is rather unclear. Instead, for most people, the name XXX_SHARED means YES=shared build and NO=static build. 

But **it is "additional"** in purest sense, see my comment (from last summary) here:

* <<<As a mention, when in static mode and ```libOM*.a```  collection is generated, the additional ```libONNXMLIR.so``` is **always linked** with the ```libLLVM.so``` and ```libMLIR.so``` **shared conterparts** and **not with the static** ones, so the object will be a truly light-weight (see the first comment how it looks).>>>

> For onnx-mlir, shared or static build is not directly configurable (it's always the same as llvm-project) so I don't think we want to use a name that give people the impression that it's configurable, which means we should avoid the word SHARED in the option name.

The statement *"it's always the same as llvm-project"* not holds, here is an example:

 * Depending on how ```llvm-project``` monorepo was configured ```onnx-mlir``` will pick up that,  **BUT** if e.g. with BUILD_SHARED_LIBS=ON override ```onnx-lib``` will generate ```libOM*.so``` objects regardless ```llvm-project``` monorepo was "static" !

>so I don't think we want to use a name that give people the impression that it's configurable, which means we should avoid the word SHARED in the option name.

Reply:
> But it is "additional" in purest sense, see my comment (from last summary) here:
>
> -  <<<As a mention, when in static mode and libOM*.a collection is generated, the additional libONNXMLIR.so is always linked with the libLLVM.so and libMLIR.so shared conterparts and not with the static ones, so the object will be a truly light-weight (see the first comment how it looks).>>>

All these might be clear to you since you added the option but you cannot expect anybody just started working with onnx-mlir to figure these out by just reading "Enable ONNX-MLIR additional shared library object". When people see an option like XXX_SHARED, the common sense response is ON=shared build and OFF=static, not ON=additional and OFF=no additional.

> Depending on how llvm-project monorepo was configured onnx-mlir will pick up that, BUT if e.g. with BUILD_SHARED_LIBS=ON override onnx-lib will generate libOM*.so objects regardless llvm-project monorepo was "static" !

That's what we do not want. We want the behavior to be that onnx-mlir is always in sync with llvm-project. That's why we did not have an option to allow you to override. I'm OK with you having an option to control whether you want a single/unified .so or not. But don't change the behavior that onnx-mlir is always in sync with llvm-project in terms of shared vs static build.